### PR TITLE
core: watchdog messages to delay statup/runtime/shutdown timeout failures

### LINKED ISSUE #5868: RFE: optionally require WATCHDOG=1 messages during the service start-up phase too
### Submission type

  - [X] Request for enhancement (RFE)

### systemd version the issue has been seen with

231

### Used distribution

Fedora (but relevant everywhere)

### In case of bug report: Expected behaviour you didn't see

A mariadb.service file with Type=notify has been written for a while. Stumbling back onto an old bug https://jira.mariadb.org/browse/MDEV-9202 we have a problem that in some cases involving large synchronisation TimeoutStartSec isn't sufficient and a sane default value can't be predicted as the size of a users data can vary greatly. A disabled value 0 isn't really sane either as stalls due to programming errors are still possible.

Can the timeouts be measured from the last sd_notify message? Pushing STATUS= messages is easy enough.

### PR DESCRIPTION
**Pull request by grooverdan**
Fixes #5868:

is using timeout_stop_usec/timeout_start_usec over the watchdog_usec the right approach?

The test case for this isn't working yet however I've manually verified its operation and it prevents the timeout of startup/shutdown as expected.

### CODE REVIEW

**Comment by poettering** [comment]
Hmm, I am not sure we should alter the behaviour like this unconditionally. We probably should add a new boolean unit setting WatchdogExtendTimeout= or so. When turned on we'd do what you propose, and extend the timeout.

I also think that we should extend the timeout by whatever is the configured watchdog timeout, under the assumption that that timeout is the one to derive the WATCHDOG=1 frequency off. Because if it was the Start/Stop timeout we'd use here (like in your current patch) then the service code would have to know that during init the start timeout is the timeout to use for WATCHDOG=1 pings, during runtime its the normal watchdog timeout, and during stop the stop timeout, and that's just too complicated. hence, let's just always extend the timeouts by the watchdog timeout, and a simple unified logic can be implemented in daemons.

Also, I think this also should be useful to extend the RuntimeMaxSec= timeout, not just the start/stop timeouts.

**Comment by poettering** [comment]
Hmm, instead of adding a unit file boolean for this, another option would be to extend the notify messages a bit. i.e. sending "WATCHDOG=1" together with "EXTEND_TIMEOUT=1" would have the effect of extending any start/stop/runtime timeout currently set in the service by the watchdog timeout.

i think I like this approach better I must say, and we wouldn't have to add yet another unit file setting to make this work.

Does that make sense?

File: src/core/service.c [hunk 1]
```diff
@@ -253,10 +255,27 @@ static void service_start_watchdog(Service *s) {
...
+        usec_t timeout;
+
         assert(s);
 
         dual_timestamp_get(&s->watchdog_timestamp);
         service_start_watchdog(s);
+
+        if (IN_SET(s->state, SERVICE_START, SERVICE_START_POST)) {
+                timeout = s->timeout_start_usec;
+        } else if (IN_SET(s->state,
```
Comment (poettering) [inline src/core/service.c]: coding style nitpick: not {} around single-line if blocks, also see CODING_STYLE

File: src/core/service.c [hunk 2]
```diff
@@ -253,10 +255,27 @@ static void service_start_watchdog(Service *s) {
...
+                    SERVICE_STOP, SERVICE_STOP_SIGABRT, SERVICE_STOP_SIGTERM, SERVICE_STOP_SIGKILL, SERVICE_STOP_POST,
+                    SERVICE_FINAL_SIGTERM, SERVICE_FINAL_SIGKILL)) {
+
+                timeout = s->timeout_stop_usec;
+        } else {
+                return;
+        }
```
Comment (poettering) [inline src/core/service.c]: as above

**Comment by grooverdan** [comment]
I like this second approach. The use of WATCHDOG=1 with EXTEND_TIMEOUT=1 makes me wonder why WATCHDOG=1 is needed (for the purposes on this RFE), EXTEND_TIMEOUT=1 seems sufficient.

Is this still resulting a Restart=on-watchdog failure? Or is a new one needed?

Runtime timeout and nits noted. I will resubmit.

**Comment by poettering** [comment]
> I like this second approach. The use of WATCHDOG=1 with EXTEND_TIMEOUT=1 makes me wonder why WATCHDOG=1 is needed (for the purposes on this RFE), EXTEND_TIMEOUT=1 seems sufficient.

Hmm, yeah, we should probably not insist on WATCHDOG=1 being set when honouring EXTEND_TIMEOUT=1. However, we should permit services to send both in combination and we should do the right thing then (i.e. consider it both a watchdog ping, as well as a request to extend any start/stop/runtime timeout that is pending). And we should document that this works. Why? because that means that service code (in particular generic event loops such as sd-event) can be made to send both in combination if desired, without having to track the precise state the service is running int

**Comment by poettering** [comment]
> Is this still resulting a Restart=on-watchdog failure? Or is a new one needed?

I don't think Restart= needs any special handling beyond what is already there at this time. I mean, we already have cocnepts for dealing with timeouts and the fact whether they are extend or not should not matter.

**Comment by grooverdan** [comment]
Implemented. With working test suite. Small change in behaviour for WATCHDOG_USEC= in that this doesn't start the watch dog service due to its now dual meaning in determining the behaviour of EXTEND_TIMEOUT=1.

Feedback welcome.

nit: seems to be ambiguous if cancelled or canceled is the correct spelling in the code.

File: src/core/service.c [hunk 3]
```diff
@@ -252,6 +254,30 @@ static void service_start_watchdog(Service *s) {
...
+        if (IN_SET(s->state, SERVICE_START, SERVICE_START_POST,
+                    SERVICE_STOP, SERVICE_STOP_SIGABRT, SERVICE_STOP_SIGTERM, SERVICE_STOP_SIGKILL, SERVICE_STOP_POST,
+                    SERVICE_FINAL_SIGTERM, SERVICE_FINAL_SIGKILL)) {
+                usec_t watchdog_usec, new_timeout;
+
+                watchdog_usec = service_get_watchdog_usec(s);
+                new_timeout = usec_add(now(CLOCK_MONOTONIC), watchdog_usec);
+                r = service_arm_timer(s, new_timeout);
+                if (r < 0)
+                    log_unit_warning_errno(UNIT(s), r, "Failed to add to shutdown/startup timer: %m");
```
Comment (poettering) [inline src/core/service.c]: strange indenting. please stick to multiples of 8ch. Alsoe see CODING_STYLE

File: src/core/service.c [hunk 4]
```diff
@@ -252,6 +254,30 @@ static void service_start_watchdog(Service *s) {
...
+
+                if (new_timeout > UNIT(s)->job_timeout) {
+                    r= service_arm_timer(s, new_timeout);
```
Comment (poettering) [inline src/core/service.c]: 8ch indenting... also space between `r` and `=`

File: src/core/service.c [hunk 5]
```diff
@@ -252,6 +254,30 @@ static void service_start_watchdog(Service *s) {
...
+                    if (r < 0)
+                        log_unit_warning_errno(UNIT(s), r, "Failed to add to runtime timer: %m");
```
Comment (poettering) [inline src/core/service.c]: 8ch indenting

File: src/core/service.c [hunk 6]
```diff
@@ -252,6 +254,30 @@ static void service_start_watchdog(Service *s) {
...
+        assert(s);
+
+        if (IN_SET(s->state, SERVICE_START, SERVICE_START_POST,
+                    SERVICE_STOP, SERVICE_STOP_SIGABRT, SERVICE_STOP_SIGTERM, SERVICE_STOP_SIGKILL, SERVICE_STOP_POST,
+                    SERVICE_FINAL_SIGTERM, SERVICE_FINAL_SIGKILL)) {
+                usec_t watchdog_usec, new_timeout;
+
+                watchdog_usec = service_get_watchdog_usec(s);
+                new_timeout = usec_add(now(CLOCK_MONOTONIC), watchdog_usec);
+                r = service_arm_timer(s, new_timeout);
```
Comment (poettering) [inline src/core/service.c]: hmm, this way as soon as we get EXTEND_TIMEOUT=1 we completely forget about any timeout set with TimeoutStartSec= and TimeoutStopSec=. I find that quite surprising I think, no? 

Instead of calling service_arm_timer() here, I'd instead do something like this:

```
if (s->timer_event_source) {
        uint64_t current, extended;
        r = sd_event_source_get_time(s->timer_event_source, &current);
        if (r < 0)
                …
        extended = usec_add(now(CLOCK_MONOTONIC), service_get_watchdog_usec(s));
        if (extended > current) {
                r = sd_event_source_set_time(s->timer_event_source, extended),
                if (r < 0) 
                        …
         }
}
```

You get the idea: we bump the timeout only if we'd actually increase it, and otherwise just leave in what's already set.

(and similar for the watchdog timeout event source)

Comment (grooverdan) [reply src/core/service.c]: Forgot to update documentation. Will look tomorrow. Need to do "similar for watchdog timeout event source too" (with tests)

File: man/sd_notify.xml [hunk 7]
```diff
@@ -202,6 +202,14 @@
         watchdog is enabled. </para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term>EXTEND_TIMEOUT=1</term>
+
+        <listitem><para>Tells the service manager to set the startup or shutdown service timeout be the watchdog time
```
Comment (keszybz) [inline man/sd_notify.xml]: Please drop the "the" in "the watchdog time".

File: man/sd_notify.xml [hunk 8]
```diff
@@ -202,6 +202,14 @@
...
+        interval in the future. The service RuntimeMaxSec is also extended if set.
```
Comment (keszybz) [inline man/sd_notify.xml]: "The maximum time the service is allowed to run (set by RuntimeMaxSec=) is also extended." (The way this was phrased implied that RuntimeMaxSec itself is extended, which I hope is not true.)

File: man/systemd.service.xml [hunk 9]
```diff
@@ -513,6 +513,12 @@
         <varname>Type=oneshot</varname> is used, in which case the
         timeout is disabled by default (see
         <citerefentry><refentrytitle>systemd-system.conf</refentrytitle><manvolnum>5</manvolnum></citerefentry>).
+        </para>
+
+        <para><varname>Type=notify</varname> services may by sending <varname>EXTEND_TIMEOUT=1</varname> cause
+        this value to be ignored after the first receipt of this message and will thereafter rely on less that
+        watchdog interval repeats of this message until the service is started (see
+        <citerefentry><refentrytitle>sd_notify</refentrytitle><manvolnum>3</manvolnum></citerefentry>).
```
Comment (keszybz) [inline man/systemd.service.xml]: This *is* grammatically correct, but somehow hard to parse. Maybe
> If a service of Type=notify sends an EXTEND_TIMEOUT=1 notification, which it may do more than once, the deadline for startup is replaced. The new deadline is watchdog interval (see WatchdogIntervalSec=) after the reception of that notification.

File: src/core/service.c [hunk 10]
```diff
@@ -252,6 +252,26 @@ static void service_start_watchdog(Service *s) {
                 log_unit_warning_errno(UNIT(s), r, "Failed to install watchdog timer: %m");
 }
 
+static void service_extend_timeout(Service *s) {
+        assert(s);
+
+        if (s->timer_event_source) {
+                uint64_t current, extended;
```
Comment (keszybz) [inline src/core/service.c]: `current` needs to be initalized to 0, otherwise it may be used uninitialized.

File: test/TEST-16-EXTEND-TIMEOUT/assess.sh [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+exit=0
+
+rm -f /test.log
+
+# This checks all stages, start, runtime and stop, can be extended by
+# EXTEND_TIMEOUT=1
+if [[ ! -f /success_all.success ]]
+then
+  journalctl -u testsuite-success-all.service >> /test.log
```
Comment (keszybz) [inline test/TEST-16-EXTEND-TIMEOUT/assess.sh]: We seem to use 4 spaces of indentiation in `test/*/test.sh`. Please make it the same here for consistency.

File: test/TEST-16-EXTEND-TIMEOUT/assess.sh [hunk 12]
```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
+set -v -x
+
+# enough time for testsuite-success to complete
+sleep 18
```
Comment (keszybz) [inline test/TEST-16-EXTEND-TIMEOUT/assess.sh]: Just make it "big", like 90s. Various delays happen.

File: test/TEST-16-EXTEND-TIMEOUT/test.sh [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+. $TEST_BASE_DIR/test-functions
+
+test_setup() {
+    create_empty_image
+    mkdir -p $TESTDIR/root
+    mount ${LOOPDEV}p1 $TESTDIR/root
+
+    # Create what will eventually be our root filesystem onto an overlay
+    (
+        LOG_LEVEL=5
```
Comment (keszybz) [inline test/TEST-16-EXTEND-TIMEOUT/test.sh]: Is this used anywhere?

Comment (grooverdan) [reply test/TEST-16-EXTEND-TIMEOUT/test.sh]: Probably not. Copied from another test. Will remove or add tests that use it. Other points noted too.

File: man/sd_notify.xml [hunk 14]
```diff
@@ -202,6 +202,14 @@
         watchdog is enabled. </para></listitem>
       </varlistentry>
 
+      <varlistentry>
```
Comment (poettering) [inline man/sd_notify.xml]: btw, we don't use Signed-off-by: on systemd. That's a kernel thing. Please drop from the commit msg

File: man/sd_notify.xml (hunk 7 above)
Comment (poettering) [reply man/sd_notify.xml]: "… to extend the start or stop service timeout by the watchdog timeout."

File: man/sd_notify.xml (hunk 8 above)
Comment (poettering) [reply man/sd_notify.xml]: also, please always enclose settings in `<varname></varname>`, and suffix them with `=`. i.e. `<varname>RuntimeMaxSec=</varname>`

File: src/core/service.c [hunk 15]
```diff
@@ -252,6 +252,26 @@ static void service_start_watchdog(Service *s) {
...
+                int r;
+
+                r = sd_event_source_get_time(s->timer_event_source, &current);
+                if (r < 0)
+                        log_unit_warning_errno(UNIT(s), r, "Failed to retrieve timeout timer: %m");
```
Comment (poettering) [inline src/core/service.c]: hmm, i'd just fail this function if sd_event_source_get_time returns < 0. Use log_unit_error_errno() and return r. This is not going to happen anyway, and if it does, then there's something seriously wrong and it's probably better to not restart the timeouts with current == 0.

File: man/sd_notify.xml (hunk 8 above)
Comment (grooverdan) [reply man/sd_notify.xml]: @keszybz - "RuntimeMaxSec itself is extended, which I hope is not true" - this is currently the case. @poettering asked for this in the https://github.com/systemd/systemd/pull/7214#issuecomment-343451927 above. Why isn't this desirable?

**Comment by grooverdan** [comment]
Is it OK to change the behaviour of https://www.freedesktop.org/software/systemd/man/systemd.service.html#WatchDogSec to start the watchdog checks after the first WATCHDOG=1 is received?

Alternately EXTEND_TIMEOUT=(usec)

File: man/sd_notify.xml (hunk 8 above)
Comment (keszybz) [reply man/sd_notify.xml]: I wasn't clear. What I meant is that the configuration setting itself should not be changed. And indeed, it's not. Only the timestamp in memory is changed, and the unit itself of its representation on disk is not changed. (We have various code which will update the unit itself, for example when properties are set over dbus, new configuration snippets are written to the file system. This is not the case here).

Comment (grooverdan) [reply man/sd_notify.xml]: Ok. makes sense. Reworded a lot and changed the implication that settings are changed, only behaviour.

File: test/TEST-16-EXTEND-TIMEOUT/assess.sh (hunk 12 above)
Comment (grooverdan) [reply test/TEST-16-EXTEND-TIMEOUT/assess.sh]: test cases have been reworked a lot. The execution of this assess.sh is just waiting on a few Wants= services to start (which are largely sleep statements). I observed a timeout of ExecStartPre=/assess.sh at 90 seconds. The sentiment however is taken and 45 seconds is used to see if any services fail. I extended the other *services to take longer and have a larger margin for scheduling. Probably not to 90s level but maybe 2000s.

**Comment by grooverdan** [comment]
Hopefully addressed all comments, not broken any more coding standards, and generally created tested working code. Rebased on master and thanks for leaving the TEST-16 namespace.

I hope the change in behaviour in bcc002d is acceptable. My rational was that Watchdog services will send WATCHDOG=1 quickly and regularly enough. There is a loss of watchdog failure if there is never a WATCHDOG=1 message sent which will now timeout rather than watchdog abort.

Thanks for your prompt reviews.

**Comment by poettering** [comment]
I am not sure I grok why you changed the watchdog logic there? Can you elaborate?

**Comment by grooverdan** [comment]
If WatchDogSec is set for the service for the purpose of using the service using EXTEND_TIMEOUT=1 the line 1070, 1777 would start the watchdog event service. Likewise WATCHDOG_USEC=X will start the service in service_reset_watchdog_timeout (line 268) even though its intended use may only be EXTENDED_TIMEOUT. This enabling of the watch event loop will cut short the runtime of the service if no WATCHDOG=1 was implemented during the service runtime.

Rather than defining constraints for service implementers around disabling the WATCHDOG_USEC=0 once EXTEND_TIMEOUT=1 has finished, or requiring watchdog to be implemented as the same time as EXTEND_TIMEOUT, I thought it was less invasive to change the existing behaviour. I did this because watchdog services will issued WATCHDOG=1 at a point at which they are in a stable loop where progress is monitored. The difference from the start of running READY=1 this point is assumed to be stable.

The alternative is to decouple the extend timeout from watchdog with either a different variable or EXTEND_TIMEOUT={usec}. I can change to this approach if desired.

As an illustration of this without this first commit a number of test cases in the second commit would fail, example testsuite-success-{all,stop}.

**Comment by grooverdan** [comment]
is this explanation sufficient?

Would you rather EXTEND_TIMEOUT be decoupled from watchdog intervals? If so, by which approach?

**Comment by keszybz** [comment]
That's a bummer with the initial watchdog timeout. I think we cannot do the change in https://github.com/systemd/systemd/pull/7214/commits/f1bfe7fd44edaef8b915d61ae16589d2bc43ea71 because that opens a significant loophole: if the service hangs between reaching ready state and the first watchdog notification, the watchdog will not trigger.

> Would you rather EXTEND_TIMEOUT be decoupled from watchdog intervals? If so, by which approach?

Yeah, I think that's the way to go. Can we just change EXTEND_TIMEOUT= to take a value in seconds (EXTEND_TIMEOUT=90)? Then every message would bump the timeout by some number.  For the caller this would not be any more complicated because they can used a fixed number, and for us the implementation would also be very simple. @poettering ?

**Comment by poettering** [comment]
@keszybz yupp, i had the very same idea the other day: we should make this EXTEND_TIMEOUT_USEC= and just make it accept a time value, and then apply it to all pending timeouts we keep for the service

**Comment by poettering** [comment]
ah, please always also leave a short message here when you force push a new version. github sometimes doesn't notify us about force pushed versions.

Patch looks good to me now, but it needs a rebase to be mergable? Any chance you can quickly do that? will merge then.

**Comment by grooverdan** [comment]
rebased. Will add comment next submission on force push. Thanks @poettering , @keszybz for your reviews and guidance.
